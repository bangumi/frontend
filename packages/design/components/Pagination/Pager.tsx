import React, { FC } from 'react'
import classNames from 'classnames'

export interface PagerProps {
  page: number
  active: boolean
  onClick: (p: number) => void
}

const Pager: FC<PagerProps> = (props) => {
  const prefixCls = 'bgm-pagination-pager'
  const actualClassnames = classNames(prefixCls, `${prefixCls}--${props.page}`, {
    [`${prefixCls}--active`]: props.active,
    [`${prefixCls}--disabled`]: !props.page
  })

  const handleClick = (): void => {
    props.onClick(props.page)
  }

  return (
    <li
      data-testid="pagination-pager"
      title={`${props.page}`}
      className={actualClassnames}
      onClick={handleClick}
      tabIndex={0}
    >
      <a rel="nofollow">{props.page}</a>
    </li>
  )
}

export default Pager
