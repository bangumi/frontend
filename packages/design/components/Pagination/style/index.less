@import '../../../theme/base';

.bgm-pagination {
  display: flex;

  &-prev {
    margin-right: 30px;
  }

  &-next {
    margin-left: 30px;
  }

  &-prev,
  &-next {
    user-select: none;
    cursor: pointer;
    height: 34px;
    width: 80px;
    border-radius: 17px;
    border: solid 2px @gray-10;

    &:hover {
      background-color: @primary-color;
      border-color: @primary-color;

      .bgm-pagination-icon {
        fill: #fff;
      }
    }

    &--hide {
      visibility: hidden;
    }
  }

  &-icon {
    fill: @gray-60;
  }

  &-pager {
    user-select: none;
    width: 34px;
    height: 34px;
    border-radius: 34px;
    border: solid 2px @gray-10;
    color: @gray-60;
    cursor: pointer;

    &--active,
    &:hover {
      color: #fff;
      background-color: @primary-color;
      border-color: @primary-color;
    }

    &--active {
      cursor: default;
    }

    & + & {
      margin-left: 10px;
    }
  }

  &-prev,
  &-next,
  &-pager {
    display: flex;
    align-items: center;
    justify-content: center;
  }
}
