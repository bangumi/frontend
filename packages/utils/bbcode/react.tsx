import React from 'react';

import { convert } from './convert';
import { Parser } from './parser';
import type { CodeNodeTypes, NodeTypes, VNode } from './types';

/* eslint-disable no-useless-escape */
const toCamelCase = (s: string) => s.replace(/(\-\w)/g, (k) => k[1]?.toUpperCase() ?? '');

const joinClassName = (cls: string | string[]) => {
  if (Array.isArray(cls)) {
    return cls.join(' ');
  }
  return cls;
};

const convertToReactProps = (node: VNode, key?: React.Key) => {
  const { props, style, className } = node;

  const transformedStyle: Record<string, string> = style
    ? Object.entries(style).reduce((pre, [key, val]) => {
        return {
          ...pre,
          [toCamelCase(key)]: val,
        };
      }, {})
    : {};

  const transformedClassName = className !== undefined ? joinClassName(className) : undefined;

  const reactProps = {
    ...props,
    style: transformedStyle,
    className: transformedClassName,
    key,
  };

  return reactProps;
};

export const renderNode = (node: NodeTypes, key?: React.Key): React.ReactElement | string => {
  if (typeof node === 'string') {
    const splittedStr: Array<string | React.ReactElement> = node.split(/\n/g);

    if (splittedStr.length < 2) {
      return splittedStr[0] ?? '';
    }
    const children = splittedStr.reduce<Array<string | React.ReactElement>>((pre, curr, idx) => {
      if (idx > 0) {
        return [...pre, React.createElement('br', null), curr];
      }
      return [...pre, curr];
    }, []);
    return React.createElement(React.Fragment, { key }, ...children);
  }

  const { type, children } = node;

  if (children) {
    return React.createElement(
      type,
      convertToReactProps(node, key),
      ...children.map((child, idx) => renderNode(child, idx)),
    );
  }
  return React.createElement(type, convertToReactProps(node, key));
};

export const render = (rawStr: string): Array<React.ReactElement | string> => {
  const nodes: CodeNodeTypes[] = new Parser(rawStr).parse();

  return nodes.map((node, idx) => renderNode(convert(node, {}), idx));
};
