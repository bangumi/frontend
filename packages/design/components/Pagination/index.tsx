import React, { FC, useState } from 'react';
import Pager from './Pager';
import { VerticalLeft, VerticalRight } from '@bangumi/icons';
import cn from 'classnames';

export interface PaginationProps {
  /* 当前偏移 */
  currentPage?: number;
  /* 单页的数据条数 */
  pageSize?: number;
  /* 数据的总条数 */
  total?: number;
  /* 页码改变的回调 */
  onChange?: (offset: number) => void;
}

function calculatePage(pageSize: number, total: number): number {
  return Math.floor((total - 1) / pageSize) + 1;
}

const Pagination: FC<PaginationProps> = ({
  currentPage = 1,
  pageSize = 30,
  total = 0,
  ...restProps
}) => {
  const [current, setCurrent] = useState(() => currentPage);

  // 不需要分页的时候不渲染
  if (total < pageSize) {
    return null;
  }

  const allPages = calculatePage(pageSize, total);
  const handleChange = (page: number): void => {
    setCurrent(page);
    restProps.onChange?.(page);
  };

  // prev, next
  const hasPrev = current > 1;
  const prev = (): void => {
    if (hasPrev) {
      handleChange(current - 1);
    }
  };
  const hasNext = current < allPages;
  const next = (): void => {
    if (hasNext) {
      handleChange(current + 1);
    }
  };

  const prevButton = (
    <li
      onClick={prev}
      data-testid='pagination-prev'
      className={cn('bgm-pagination-prev', {
        'bgm-pagination-prev--hide': !hasPrev,
      })}
    >
      <VerticalLeft className='bgm-pagination-icon' />
    </li>
  );

  const nextButton = (
    <li
      onClick={next}
      data-testid='pagination-next'
      className={cn('bgm-pagination-next', {
        'bgm-pagination-next--hide': !hasNext,
      })}
    >
      <VerticalRight className='bgm-pagination-icon' />
    </li>
  );

  // pagers
  const pagerList = [];
  const pagerProps = {
    onClick: handleChange,
  };
  const L_PAGE_BUFFER_SIZE = 3;
  const R_PAGE_BUFFER_SIZE = 6;
  const PAGE_BUFFER_SIZE = L_PAGE_BUFFER_SIZE + R_PAGE_BUFFER_SIZE;
  if (allPages <= PAGE_BUFFER_SIZE) {
    for (let i = 1; i <= allPages; i += 1) {
      const active = current === i;
      pagerList.push(<Pager {...pagerProps} key={i} page={i} active={active} />);
    }
  } else {
    let left = Math.max(1, current - L_PAGE_BUFFER_SIZE);
    let right = Math.min(current + R_PAGE_BUFFER_SIZE, allPages);

    if (current - 1 < L_PAGE_BUFFER_SIZE) {
      right = PAGE_BUFFER_SIZE + 1;
    }
    if (allPages - current <= R_PAGE_BUFFER_SIZE) {
      left = allPages - PAGE_BUFFER_SIZE;
    }

    for (let i = left; i <= right; i += 1) {
      const active = current === i;
      pagerList.push(<Pager {...pagerProps} key={i} page={i} active={active} />);
    }
  }
  return (
    <ul className='bgm-pagination' data-testid='pagination-wrapper'>
      {prevButton}
      {pagerList}
      {nextButton}
    </ul>
  );
};

export default Pagination;
