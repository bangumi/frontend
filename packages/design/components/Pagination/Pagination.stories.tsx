import React from 'react'
import Pagination from '.'
import { ComponentStory, ComponentMeta } from '@storybook/react'

const componentMeta: ComponentMeta<typeof Pagination> = {
  title: 'modern/Pagination',
  component: Pagination
}

export default componentMeta

const Template: ComponentStory<typeof Pagination> = (args) => {
  return <Pagination {...args} />
}

export const Usage = Template.bind({})
Usage.args = {
  total: 3000,
  currentOffset: 20
}
