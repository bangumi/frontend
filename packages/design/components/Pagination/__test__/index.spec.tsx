import { fireEvent, render, waitFor } from '@testing-library/react';
import React from 'react';

import Pagination from '..';

it('should render correctly', () => {
  const { getByTestId } = render(<Pagination total={3939} />);
  expect(getByTestId('pagination-wrapper')).toMatchSnapshot();
});

it('should hightlight current page and not highlight other page', async () => {
  const currentPage = 20;
  const { findAllByTestId } = render(<Pagination total={3000} currentPage={currentPage} />);
  const pagers = await findAllByTestId('pagination-pager');
  expect(pagers.length).toBe(10);
  pagers.forEach((pager) => {
    if (pager.title === String(currentPage)) {
      expect(pager.classList.contains('bgm-pagination-pager--active')).toBe(true);
    } else {
      expect(pager.classList.contains('bgm-pagination-pager--active')).toBe(false);
    }
  });
});

it('should response mouse click right', async () => {
  const onChange = vi.fn();
  const { findAllByTestId } = render(
    <Pagination total={30} pageSize={10} currentPage={1} onChange={onChange} />,
  );
  const pagers = await findAllByTestId('pagination-pager');
  expect(pagers.length).toBe(3);

  const page2 = pagers.at(1)!;
  expect(page2.classList.contains('bgm-pagination-pager--2')).toBe(true);

  fireEvent.click(page2);
  expect(page2.classList.contains('bgm-pagination-pager--active')).toBe(true);
  expect(onChange).toHaveBeenLastCalledWith(2);
});

it('should not response the prev-button clicked', () => {
  const onChange = vi.fn();
  const { queryByTestId } = render(<Pagination total={3000} currentPage={1} onChange={onChange} />);
  const prevButton = queryByTestId('pagination-prev')!;
  fireEvent.click(prevButton);
  expect(onChange).toHaveBeenCalledTimes(0);
  const nextButton = queryByTestId('pagination-next')!;
  fireEvent.click(nextButton);
  expect(onChange).toHaveBeenLastCalledWith(2);
});
it('should not response the next-button clicked', () => {
  const onChange = vi.fn();
  const { queryByTestId } = render(
    <Pagination total={3000} currentPage={100} onChange={onChange} />,
  );
  const nextButton = queryByTestId('pagination-next')!;
  fireEvent.click(nextButton);
  expect(onChange).toHaveBeenCalledTimes(0);
  const prevButton = queryByTestId('pagination-prev')!;
  fireEvent.click(prevButton);
  expect(onChange).toHaveBeenLastCalledWith(99);
});

it('should response next page', async () => {
  const onChange = vi.fn();
  const { getByTestId } = render(<Pagination total={300} onChange={onChange} />);
  const nextButton = getByTestId('pagination-next');
  fireEvent.click(nextButton);
  expect(onChange).toHaveBeenLastCalledWith(2);

  await waitFor(() => {
    const prevButton = getByTestId('pagination-prev');
    fireEvent.click(prevButton);
    expect(onChange).toHaveBeenLastCalledWith(1);
  });
});
