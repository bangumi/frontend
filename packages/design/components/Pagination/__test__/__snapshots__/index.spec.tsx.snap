// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render correctly 1`] = `
<ul
  class="bgm-pagination"
  data-testid="pagination-wrapper"
>
  <li
    class="bgm-pagination-prev bgm-pagination-prev--hide"
    data-testid="pagination-prev"
  >
    <div
      class="bgm-pagination-icon"
    />
  </li>
  <li
    class="bgm-pagination-pager bgm-pagination-pager--1 bgm-pagination-pager--active"
    data-testid="pagination-pager"
    tabindex="0"
    title="1"
  >
    <a
      rel="nofollow"
    >
      1
    </a>
  </li>
  <li
    class="bgm-pagination-pager bgm-pagination-pager--2"
    data-testid="pagination-pager"
    tabindex="0"
    title="2"
  >
    <a
      rel="nofollow"
    >
      2
    </a>
  </li>
  <li
    class="bgm-pagination-pager bgm-pagination-pager--3"
    data-testid="pagination-pager"
    tabindex="0"
    title="3"
  >
    <a
      rel="nofollow"
    >
      3
    </a>
  </li>
  <li
    class="bgm-pagination-pager bgm-pagination-pager--4"
    data-testid="pagination-pager"
    tabindex="0"
    title="4"
  >
    <a
      rel="nofollow"
    >
      4
    </a>
  </li>
  <li
    class="bgm-pagination-pager bgm-pagination-pager--5"
    data-testid="pagination-pager"
    tabindex="0"
    title="5"
  >
    <a
      rel="nofollow"
    >
      5
    </a>
  </li>
  <li
    class="bgm-pagination-pager bgm-pagination-pager--6"
    data-testid="pagination-pager"
    tabindex="0"
    title="6"
  >
    <a
      rel="nofollow"
    >
      6
    </a>
  </li>
  <li
    class="bgm-pagination-pager bgm-pagination-pager--7"
    data-testid="pagination-pager"
    tabindex="0"
    title="7"
  >
    <a
      rel="nofollow"
    >
      7
    </a>
  </li>
  <li
    class="bgm-pagination-pager bgm-pagination-pager--8"
    data-testid="pagination-pager"
    tabindex="0"
    title="8"
  >
    <a
      rel="nofollow"
    >
      8
    </a>
  </li>
  <li
    class="bgm-pagination-pager bgm-pagination-pager--9"
    data-testid="pagination-pager"
    tabindex="0"
    title="9"
  >
    <a
      rel="nofollow"
    >
      9
    </a>
  </li>
  <li
    class="bgm-pagination-pager bgm-pagination-pager--10"
    data-testid="pagination-pager"
    tabindex="0"
    title="10"
  >
    <a
      rel="nofollow"
    >
      10
    </a>
  </li>
  <li
    class="bgm-pagination-next"
    data-testid="pagination-next"
  >
    <div
      class="bgm-pagination-icon"
    />
  </li>
</ul>
`;
